utils/initial_data_preparation.py

```diff
@@ -277,32 +316,28 @@ def get_base_repo_url(url):
         f"file and saved in {output_dir} "
     )
 
-    # Some of the URLs end with "/". I need to remove them.
-    df["repourl"] = df["repourl"].str.rstrip("/")
+    # Strip leading '+' characters, then remove trailing slashes, and strip
+    # leading / trailing spaces,
+    df["repourl"] = df["repourl"].str.lstrip("+").str.rstrip("/").str.strip()
 
-    # Removing duplicate rows
-    remove_duplicates(df)
-
-    # Removing null value if any (There is no null values in the data at this
-    # moment)
-    remove_null_values(df)
+    # Marking duplicate rows and Null values
+    df = mark_duplicates(df)
+    df = mark_null_values(df)
 
     # replace http with https
     df["repourl"] = df["repourl"].str.replace(r"^http\b", "https", regex=True)
 
-    # Extract the domain from the URL
-    df["repodomain"] = df["repourl"].str.extract(r"https?://(www\.)?([^/]+)")[1]
-
-    # Extracting the base repo url
-    df["base_repo_url"] = df["repourl"].apply(get_base_repo_url)
+    # Extracts domains from URLs and flags rows with unsupported URL schemes.
+    # unsupported_url_flag: A boolean flag that is True for rows where the
+    # domain extraction returned None
+    df = extract_and_flag_domains(df)
 
-    filter_out_incomplete_urls(df)
+    # Identifies incomplete URLs in the DataFrame by adding a new boolean column
+    # `incomplete_url_flag`.
+    df = filter_out_incomplete_urls(df)
```

Comment:
I find the `mark_...` function names clearer - unless we're actually removing these rows with incomplete URLs from the massive dataframe. How about renaming the function if it's now marking rather than removing rows?

Reply:
Yes. I will do this now.